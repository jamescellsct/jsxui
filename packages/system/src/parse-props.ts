import createSort from 'sort-css-media-queries/lib/create-sort'
import { flattenProp } from './flatten-prop'
import type { Transform } from './types'

// TODO: look into supporting level 4 syntax and sorting
// https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#syntax_improvements_in_level_4
const sortMediaQueries = createSort()

/**
 * Takes a props object containing media and state props and returns a CSS prop object.
 */
export function parseProps<
  //splitProps
  Props extends Record<string, unknown>,
  Transforms extends Record<string, Transform>,
  States extends Record<string, boolean>,
  Theme extends Record<'mediaQueries' | string, unknown>
>({
  props,
  transforms,
  states,
  theme,
}: {
  props: Props
  transforms: Transforms
  states: States
  theme: Theme
}) {
  const activeStateKeys = Object.entries(states || {})
    .filter(([, value]) => value)
    .map(([key]) => key)
  const lastActiveStateKey = activeStateKeys[activeStateKeys.length - 1]
  const attributes: Record<string, unknown> = {}
  const styles: Partial<Pick<typeof props, keyof Transforms>> = {}
  const parsedProps: Partial<Props> = {}

  for (const prop in props) {
    const value = props[prop]

    if (typeof value === 'object') {
      // parsedProps[prop] = value
    } else {
      const transformedValue = transforms[prop].call(null, value)
      parsedProps[prop] = transformedValue
    }
    //   parsedProps = {
    //     ...parsedProps,
    //     ...Object.fromEntries(
    //       flattenStyleProp(prop, value, transforms, states, theme)
    //     ),
    //   }
  }

  return parsedProps

  // const flattenedMediaStyles = Object.entries(parsedProps).reduce(
  //   (styles, [key, value]) => {
  //     if (styles[key]) {
  //       //@ts-ignore
  //       return { ...styles, [key]: { ...styles[key], ...value } }
  //     }
  //     return { ...styles, [key]: value }
  //   },
  //   {}
  // )

  // return Object.fromEntries(
  //   Object.entries(flattenedMediaStyles).sort((a, b) =>
  //     sortMediaQueries(a[0], b[0])
  //   )
  // )
}
