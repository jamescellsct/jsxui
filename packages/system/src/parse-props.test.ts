import { parseProps } from './parse-props'

const theme = {
  mediaQueries: {
    small: '(min-width: 0px)',
    medium: '(min-width: 720px)',
    large: '(min-width: 1280px)',
    dark: '(prefers-color-scheme: dark)',
  },
  colors: {
    foreground: { initial: '#000', dark: '#fff' },
    foregroundInteractive: 'blue',
    background: { initial: '#fff', dark: '#000' },
  },
}

const transforms = {
  color: (value) => ({ color: value }),
  fontSize: (value) => ({ fontSize: value }),
}

const variantProps = {
  as: {
    initial: 'p',
    descendant: 'span',
  },
  fontSize: {
    initial: 16,
    medium: 20,
    large: 24,
  },
  color: 'foreground',
}

const parsedVariantProps = {
  global: {
    initial: {
      '--font-size-heading1': '16px',
    },
    medium: {
      '--font-size-heading1': '24px',
    },
    large: {
      '--font-size-heading1': '32px',
    },
  },
  local: {
    fontSize: 'var(--font-size-heading1)',
    color: 'var(--color-foreground)',
  },
}

test('parses media and state props', () => {
  const parsedProps = parseProps({
    props: variantProps,
    states: { descendant: false, hover: false },
    transforms,
    theme,
  })

  expect(parsedProps).toMatchSnapshot()
})
