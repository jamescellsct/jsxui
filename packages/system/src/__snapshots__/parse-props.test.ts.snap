// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses media and state props 1`] = `
Object {
  "color": "foreground",
}
`;
